### PR TITLE
Improve zsh completion performance

### PR DESCRIPTION
**Pull request by marlonrichert**
* Move main completion cache handling to `brew update`.
* Enable completion caching by default.

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Improve perceived zsh completion performance by 
* delegating most of the cache handling to when `brew update` is being run, instead of doing it when the user tries to get a completion, since the user is expecting to wait during `brew update` anyway, and by 
* enabling caching of brew completions by default, since many users don't know how to enable this feature and since it's not enabled by default for new Zsh users.

I have not written tests for this, because I have no idea how to test this. I looked at other PRs for Zsh completions, but they don't seem to have included tests either. I guess no one really know how to test them.

### CODE REVIEW

File: Library/Homebrew/cmd/update-report.rb [hunk 1]
```diff
@@ -113,7 +113,7 @@ def update_report
     end
 
     if !updated
-      puts "Already up-to-date." if !args.preinstall? && !ENV["HOMEBREW_UPDATE_FAILED"]
+      puts "Homebrew already up-to-date." if !args.preinstall? && !ENV["HOMEBREW_UPDATE_FAILED"]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update-report.rb:116]: Can we make this text change in another PR? Thanks!

Comment (marlonrichert) [reply Library/Homebrew/cmd/update-report.rb:116]: Why I changed it is because it looks really weird when you first get to see text about completions being updated and then text that says "Already up-to-date." That feels rather confusing.

**Comment by marlonrichert** [comment]
> Makes sense to me so far! Haven't reviewed #7767 yet but it may be worth thinking about that, too.

@MikeMcQuaid I'm not sure what that PR is trying to do and how it relates to this one, but I'm open to any changes you want to suggest.

File: Library/Homebrew/cmd/update-report.rb (hunk 1 above)
Comment (nhooyr) [reply Library/Homebrew/cmd/update-report.rb:116]: I don't think the completion being updated is important enough to show a message. It should just happen.

**Comment by MikeMcQuaid** [comment]
> @MikeMcQuaid I'm not sure what that PR is trying to do and how it relates to this one, but I'm open to any changes you want to suggest.

@marlonrichert It's a different approach. Try it out (it's been committed to `master`). I think it provides a sufficient enough speedup that this PR may now be unnecessary.

File: Library/Homebrew/cmd/update-report.rb (hunk 1 above)
Comment (marlonrichert) [reply Library/Homebrew/cmd/update-report.rb:116]: That could be an option, too.

Or I could have look at how to get it into the update report at the end.

**Comment by marlonrichert** [comment]
@MikeMcQuaid OK, thanks. I’ll try it out.

**Comment by marlonrichert** [comment]
I looked at the code of #7767 and I think I should be able to modify it to include formulae, too.

File: Library/Homebrew/cmd/update-report.rb (hunk 1 above)
Comment (marlonrichert) [reply Library/Homebrew/cmd/update-report.rb]: I removed the text output for completion updating and reverted the "up-to-date" text.

**Comment by marlonrichert** [comment]
@MikeMcQuaid I just tested `master` and although completion for Homebrew commands is now faster, there's still a slight noticeable delay. Whereas with this PR, after doing `brew update`, completion for commands, formulas and casks is instantaneous. Plus, #7767 does nothing to speed up completion for Homebrew formulas or casks.

So, I hope you will still consider merging this. #7767 has not made it obsolete.

File: Library/Homebrew/utils/update-completions.zsh [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  [[ -f "$HOMEBREW_CACHE/all_commands_list.txt" ]] &&
+    list=($(cat "$HOMEBREW_CACHE/all_commands_list.txt")) ||
+    list=($(cat "$HOMEBREW_REPOSITORY/completions/internal_commands_list.txt"))
+  list=(${list:#*instal}) # Exclude instal, uninstal, etc.
+  _store_cache $comp_cachename list
+fi
+
+comp_cachename=brew_formulae
+if _cache_invalid $comp_cachename; then
+  list=( $(brew search) )
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/update-completions.zsh:40]: What about having the output of this be cached in `HOMEBREW_CACHE` like `all_commands_list.txt` on `brew update` instead? That would enable it to be used in Bash completion too. CC @alebcay for thoughts.

Comment (marlonrichert) [reply Library/Homebrew/utils/update-completions.zsh:40]: `__brew_installed_formulae` and `__brew_installed_casks` could also be cached. Those are quite slow, too.

Comment (marlonrichert) [reply Library/Homebrew/utils/update-completions.zsh:40]: > I think ideally the formula list would be generated with Ruby and cached in `HOMEBREW_CACHE` and then consumed by whatever completions we'd like. We could still store a copy of this list in the Zsh completions cache if that makes the completion execution marginally faster.

The Zsh completion code could just read it straight out of `HOMEBREW_CACHE`, instead of using Zsh's own caching mechanism. I need to test it, but I think it should be equally fast.

Comment (alebcay) [reply Library/Homebrew/utils/update-completions.zsh:40]: > Homebrew's Zsh completion code could just read the completions straight out of HOMEBREW_CACHE

I believe this is what we're doing right now after #7767 was merged. I left the existing completion cache code in since it would still be useful for the other Zsh completions (formulae and casks), and wouldn't hurt anything.

Comment (marlonrichert) [reply Library/Homebrew/utils/update-completions.zsh:40]: Ah, you're right: https://github.com/Homebrew/brew/pull/7767/files#diff-c0df35ca56e49697ec906a8ce9a45448R151

Sorry, I missed that.

Comment (marlonrichert) [reply Library/Homebrew/utils/update-completions.zsh:40]: > What about having the output of this be cached in `HOMEBREW_CACHE` like `all_commands_list.txt` on brew update instead? That would enable it to be used in Bash completion too. 

@MikeMcQuaid I'm all for it, but can we do that as a separate PR? I would like to get this one merged now.

Comment (MikeMcQuaid) [reply Library/Homebrew/utils/update-completions.zsh:40]: @marlonrichert I'd rather not make the `cmd/update.sh` change given the discussion above, sorry. I'm open to other parts of this PR being merged. When `cmd/update.sh` fails: no-one can update Homebrew so I'm extremely reluctant to make changes there that aren't 100% necessary.

Comment (MikeMcQuaid) [reply Library/Homebrew/utils/update-completions.zsh:40]: > `__brew_installed_formulae` and `__brew_installed_casks` could also be cached. Those are quite slow, too.

This seems like a good idea. This cache could be updated on install/uninstall of these commands.

Comment (marlonrichert) [reply Library/Homebrew/utils/update-completions.zsh:40]: > I'd rather not make the `cmd/update.sh` change given the discussion above, sorry. I'm open to other parts of this PR being merged. When `cmd/update.sh` fails: no-one can update Homebrew so I'm extremely reluctant to make changes there that aren't 100% necessary.

OK, I'll remove the `cmd/update.sh` changes from this PR.

Comment (MikeMcQuaid) [reply Library/Homebrew/utils/update-completions.zsh:40]: Thanks! Please roll Library/Homebrew/utils/update-completions.zsh back into completions/zsh/_brew too. After that, though, I'm not sure what's actually changing here?

Comment (marlonrichert) [reply Library/Homebrew/utils/update-completions.zsh:40]: What’s left are three things that speed up Zsh completions:
* Enable Zsh completion caching for `brew` arguments by default. (Right now, it’s disabled by default.)
* Use `$(<file)` instead of `$(cat file). (The former is faster than the latter, because it does not spawn a new process, which the latter does.)
* Don’t call `_cache_invalid` before calling `_retrieve_cache`. (The latter already calls the former. No need to call it twice.)

Comment (MikeMcQuaid) [reply Library/Homebrew/utils/update-completions.zsh:40]: Ok, cool, sounds good 👍

Comment (marlonrichert) [reply Library/Homebrew/utils/update-completions.zsh]: @MikeMcQuaid  I made the changes. Please review.
